src/x86/sad_plane.asm

```diff
@@ -47,103 +49,132 @@ mask_lut: db \
 -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1, 0, 0, \
 -1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1, 0,
 
-SECTION .text
+jmp_table_avx2: dq \
+  mangle(private_prefix %+ _sad_plane_8bpc_avx2).vec0, \
+  mangle(private_prefix %+ _sad_plane_8bpc_avx2).vec1, \
+  mangle(private_prefix %+ _sad_plane_8bpc_avx2).vec2, \
+  mangle(private_prefix %+ _sad_plane_8bpc_avx2).vec3
 
-%if ARCH_X86_64
+jmp_table_sse2: dq \
+  mangle(private_prefix %+ _sad_plane_8bpc_sse2).vec0, \
+  mangle(private_prefix %+ _sad_plane_8bpc_sse2).vec1, \
+  mangle(private_prefix %+ _sad_plane_8bpc_sse2).vec2, \
+  mangle(private_prefix %+ _sad_plane_8bpc_sse2).vec3
 
-%macro SAD_ROW_FN 0
-cglobal sad_row_8bpc, 3, 5, 8, p1, p2, len, \
-                      resid_simd, resid
-  ; always need to zero first register
-  pxor    xm0, xm0
-  mov     resid_simdq, lenq
-  mov     residq,  lenq
+%use ifunc
+
+SECTION .text
+
+%macro SAD_PLANE_FN 0
+cglobal sad_plane_8bpc, 5, 9, 9, p1, p2, stride, width, rows, \
+                      resid_simd, resid, width_unrll, tmp0
+  mov     resid_simdq, widthq
+  mov     residd, widthd
   and     residd, mmsize - 1
   and     resid_simdq, -(mmsize)
-  jz     .residual
-  ; need to zero 2 more registers
+  and     widthq, -(4*mmsize)
+  ; LUT row size is always 32 regardless of mmsize (because the
+  ; start of the rows would be the same, so we reuse the same LUT)
+  shl     residd, ilog2(32)
+  pxor    xm0, xm0
   pxor    xm1, xm1
   pxor    xm2, xm2
-  ; len = num elements processed by unrolled loop
-  and     lenq, -(4*mmsize)
-  jz     .lt_4x
-  add     p1q, lenq
-  add     p2q, lenq
-  sub     resid_simdq, lenq
-  ; last register needed for unrolled loop
   pxor    xm3, xm3
-  neg     lenq
+  ; load mask from lookup table into m8
+  lea   tmp0q, [mask_lut]
+  mova     m8, [tmp0q + residq]
+
+  DEFINE_ARGS p1, p2, stride, width, rows, \
+                      resid_simd, resid, width_unrll, skip_ptr
+
+  sub     resid_simdq, widthq
+  ; need to divide by mmsize to load skip pointer
+  shr     resid_simdq, ilog2(mmsize)
+%if mmsize == 32
+  %define jmp_table jmp_table_avx2
+%elif mmsize == 16
+  %define jmp_table jmp_table_sse2
+%endif
+  lea     skip_ptrq, [jmp_table]
+  mov     skip_ptrq, [skip_ptrq + 8*resid_simdq]
+  ; shift back (for residual to load correct number of bytes)
+  shl     resid_simdq, ilog2(mmsize)
+  ; set pointer to point after end of width of first row
+  add     p1q, widthq
+  add     p2q, widthq
+  mov     width_unrllq, widthq
+  neg     widthq
+.loop_setup:
```

Comment:
maybe loop_row would make more sense as a label?

Reply:
Seems reasonable, I changed the label to `loop_row`